File: examples/python/best-practices/custom-serialization.py
from dataclasses import dataclass, asdict

@dataclass
class Order:
    order_id: str
    total: float
    items: list

@durable_step
def process_order(step_context: StepContext, order_data: dict) -> dict:
    order = Order(**order_data)
    # Process order...
    return asdict(order)  # Convert dataclass to dict

File: examples/java/best-practices/custom-serialization.java
// Use a POJO with Jackson-compatible annotations
public class Order {
    private String orderId;
    private double total;
    private List<String> items;
    // getters, setters, no-arg constructor
}

var order = ctx.step("process_order", Order.class, stepCtx -> {
    var o = new Order();
    o.setOrderId(orderData.get("order_id"));
    o.setTotal((double) orderData.get("total"));
    o.setItems((List<String>) orderData.get("items"));
    // Process order...
    return o;
});
